### PR TITLE
Update the hazelcast discovery to use the service account and https.

### PR DESCRIPTION
**Pull request by brendandburns**
See https://github.com/GoogleCloudPlatform/kubernetes/issues/7444

### CODE REVIEW

File: src/main/java/com/github/pires/hazelcast/HazelcastDiscoveryController.java [hunk 1]
```diff
@@ -43,22 +55,24 @@
...
   }
 
   @JsonIgnoreProperties(ignoreUnknown = true)
   static class Endpoints {
-
     public List<Subset> subsets;
   }
 
+  private static String getServiceAccountToken() throws IOException {
+    String file = "/var/run/secrets/kubernetes.io/serviceaccount/token";
```
Comment (pires) [inline src/main/java/com/github/pires/hazelcast/HazelcastDiscoveryController.java]: How is this file provisioned?

File: src/main/java/com/github/pires/hazelcast/HazelcastDiscoveryController.java [hunk 2]
```diff
@@ -66,23 +80,43 @@ private static String getEnvOrDefault(String var, String def) {
...
   @Override
   public void run(String... args) {
-    final String hostName = getEnvOrDefault("KUBERNETES_RO_SERVICE_HOST",
-        "localhost");
-    final String hostPort = getEnvOrDefault("KUBERNETES_RO_SERVICE_PORT",
-        "8080");
     String serviceName = getEnvOrDefault("HAZELCAST_SERVICE", "hazelcast");
     String path = "/api/v1beta3/namespaces/default/endpoints/";
-    final String host = "http://" + hostName + ":" + hostPort;
+    final String host = "https://kubernetes.default.cluster.local";
```
Comment (pires) [inline src/main/java/com/github/pires/hazelcast/HazelcastDiscoveryController.java]: Last time I checked kubernetes.local was the default domain. Has this changed? 

Also, what happens if the user specifies a different domain?

File: src/main/java/com/github/pires/hazelcast/HazelcastDiscoveryController.java (hunk 1 above)
Comment (brendandburns) [inline src/main/java/com/github/pires/hazelcast/HazelcastDiscoveryController.java]: It's automatically included in all pods now.  See:

https://github.com/GoogleCloudPlatform/kubernetes/pull/7101

File: src/main/java/com/github/pires/hazelcast/HazelcastDiscoveryController.java (hunk 2 above)
Comment (brendandburns) [inline src/main/java/com/github/pires/hazelcast/HazelcastDiscoveryController.java]: Yeah, it was changed in:

https://github.com/GoogleCloudPlatform/kubernetes/pull/8105

I suppose we could customize it, but given the auth stuff, it's going to be fairly complicated to get it all working correctly.

File: src/main/java/com/github/pires/hazelcast/HazelcastDiscoveryController.java [hunk 3]
```diff
@@ -66,23 +80,43 @@ private static String getEnvOrDefault(String var, String def) {
...
     try {
+      String token = getServiceAccountToken();
+
+      SSLContext ctx = SSLContext.getInstance("SSL");
+      ctx.init(null, trustAll, new SecureRandom());
+
       URL url = new URL(host + path + serviceName);
+      HttpsURLConnection conn = (HttpsURLConnection)url.openConnection();
+      conn.setSSLSocketFactory(ctx.getSocketFactory());
+      conn.setHostnameVerifier(trustAllHosts);
```
Comment (liggitt) [inline src/main/java/com/github/pires/hazelcast/HazelcastDiscoveryController.java]: TODO to remove once a CA is available

Comment (brendandburns) [inline src/main/java/com/github/pires/hazelcast/HazelcastDiscoveryController.java]: done.

**Comment by brendandburns** [comment]
Comments addressed.  Please re-check.

Thanks
--brendan

**Comment by pires** [comment]
Thank you @brendandburns

File: src/main/java/com/github/pires/hazelcast/HazelcastDiscoveryController.java (hunk 1 above)
Comment (pires) [inline src/main/java/com/github/pires/hazelcast/HazelcastDiscoveryController.java]: Is this present in 0.17.0? Cause
```
015-05-20T11:41:15.444282980Z 2015-05-20 11:41:15.439  INFO 12 --- [           main] c.g.p.h.HazelcastDiscoveryController     : Asking k8s registry at https://kubernetes.default.cluster.local..
2015-05-20T11:41:15.447430811Z 2015-05-20 11:41:15.447  WARN 12 --- [           main] c.g.p.h.HazelcastDiscoveryController     : Request to Kubernetes API failed
2015-05-20T11:41:15.447455987Z
2015-05-20T11:41:15.447459540Z java.nio.file.NoSuchFileException: /var/run/secrets/kubernetes.io/serviceaccount/token
2015-05-20T11:41:15.447462767Z 	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:86)
2015-05-20T11:41:15.447465764Z 	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
2015-05-20T11:41:15.447468341Z 	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
2015-05-20T11:41:15.447470841Z 	at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:214)
2015-05-20T11:41:15.447473430Z 	at java.nio.file.Files.newByteChannel(Files.java:361)
2015-05-20T11:41:15.447476757Z 	at java.nio.file.Files.newByteChannel(Files.java:407)
2015-05-20T11:41:15.447479365Z 	at java.nio.file.Files.readAllBytes(Files.java:3152)
2015-05-20T11:41:15.447481931Z 	at com.github.pires.hazelcast.HazelcastDiscoveryController.getServiceAccountToken(HazelcastDiscoveryController.java:73)
2015-05-20T11:41:15.447484868Z 	at com.github.pires.hazelcast.HazelcastDiscoveryController.run(HazelcastDiscoveryController.java:107)
2015-05-20T11:41:15.447487676Z 	at org.springframework.boot.SpringApplication.runCommandLineRunners(SpringApplication.java:672)
2015-05-20T11:41:15.447490031Z 	at org.springframework.boot.SpringApplication.afterRefresh(SpringApplication.java:690)
2015-05-20T11:41:15.447492460Z 	at org.springframework.boot.SpringApplication.run(SpringApplication.java:321)
2015-05-20T11:41:15.447494917Z 	at org.springframework.boot.builder.SpringApplicationBuilder.run(SpringApplicationBuilder.java:139)
2015-05-20T11:41:15.447497353Z 	at com.github.pires.hazelcast.Application.main(Application.java:27)
2015-05-20T11:41:15.447500453Z 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2015-05-20T11:41:15.447502947Z 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2015-05-20T11:41:15.447505408Z 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2015-05-20T11:41:15.447517981Z 	at java.lang.reflect.Method.invoke(Method.java:497)
2015-05-20T11:41:15.447520811Z 	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:53)
2015-05-20T11:41:15.447523966Z
```

Comment (pires) [inline src/main/java/com/github/pires/hazelcast/HazelcastDiscoveryController.java]: It wasn't. 0.17.1 has just been tagged. Let me give it a try.

Comment (brendandburns) [inline src/main/java/com/github/pires/hazelcast/HazelcastDiscoveryController.java]: I don't think it's in 0.17.1

Brendan

**Pull request merged**